static/js/message_edit.js

```diff
@@ -1139,6 +1139,20 @@ export function delete_topic(stream_id, topic_name) {
         data: {
             topic_name,
         },
+        success() {},
+        error(xhr) {
+            if (xhr.status === 502) {
+                /* When trying to delete a very large topic, it's
+                   possible for the request to the server to
+                   time out after making some progress. Retry the
+                   request, so that the user can just do nothing and
+                   watch the topic slowly be deleted.
+
+                   TODO: Show a nice loading indicator experience.
+                */
+                delete_topic(stream_id, topic_name);
+            }
+        },
```

Comment:
An unlimited retry load on 502s is the kind of code we don't want to have exist because you can DoS yourself. I think given the very infrequent nature of calls to this code path, it's probably fine; but in an ideal world we'd limit this to 10 retries or something.

Is it easy to add that?

Reply:
@timabbott  Yeah, it crossed my mind but the retries just run into rate limits, but not sure if that's sufficient for us? I added some code to limit the retries in the frontend now though, it's awkward but we do something similar in `static/js/server_events.js` with `get_events_failures`.

Test manually that it works in dev env:
1. Failure case, by haaving the server return 502 infinitely without doing anything. The frontend stops at 10 tries indeed.
2. Success, where the server returns 502 after processing a tiny batch. The app keeps retrying until the topic is deleted (as long as that's under 10 tries).

Reply:
I think you don't need a global variable, which is problematic because multiple of these requests can be in flight at the same time. Just pass `delete_topic_variables + 1` to `delete_topic` in the recursive call, and have the initial call assume 0 if the parameter isn't provided.

Reply:
@timabbott  Makes sense, done. Passes the same manual testing as before.

Reply:
I'm not OK making "502 means continue" into an API pattern.

If we're making the API for this endpoint include "it may not delete everything and you may need to retry" then we should document that, and add a server-side timeout _which isn't the nginx "harikari" but before it_ which causes the client to re-send the request.

502's should be exceptional cases that we aim to drive down to 0, not part of our contract with our official clients.